### ISSUE TITLE
StructuredChatAgent uses SQLDatabaseToolkit tools in wrong order

### ISSUE THREAD

**Issue by b-dulaney** [langchain-ai/langchain#7889]
### System Info

Langchain version - 0.0.235
Python version - 3.10

### Information

- [ ] The official example notebooks/scripts
- [X] My own modified scripts

### Related Components

- [ ] LLMs/Chat Models
- [ ] Embedding Models
- [ ] Prompts / Prompt Templates / Prompt Selectors
- [ ] Output Parsers
- [ ] Document Loaders
- [ ] Vector Stores / Retrievers
- [ ] Memory
- [X] Agents / Agent Executors
- [X] Tools / Toolkits
- [ ] Chains
- [ ] Callbacks/Tracing
- [ ] Async

### Reproduction

Hello,

I'm trying to use a Structured Chat agent with SQL tools as well as a VectorStoreQA tool in order to retrieve data from a Postgres database and our Pinecone store. For the Agent's chat model, I'm using Azure OpenAI's gpt-3.5-turbo version 0301.

I have tried providing the tools via the SQLDatabaseToolkit's get_tools() function as well as importing the individual tools and modifying their descriptions, but I get the same issue with both approaches. Despite the descriptions and formatting instructions provided below, the agent has a lot of trouble with querying the DB. 

It almost always starts off with 'sql_db_query' to construct a query that errors out, which then prompts the agent to check the list of tables and then their schemas. From here, it can usually arrive at the correct query - but this is not the correct behavior. Other times, it will get stuck in a loop of constructing a query, getting an error, and then either checking the query or just constructing a new one. The errors are due to the schema that it hallucinates because it didn't use the 'sql_db_list_tables' and 'sql_db_schema' tools first.

In the code excerpt below, you can see the formatting instructions provided as well as the descriptions that I'm giving the SQL tools.

Has anybody else had issues with SQL database tools and the StructuredChat agent? Should I be using a different type of agent for this?

``````
FORMAT_INSTRUCTIONS = """Use a json blob to specify a tool by providing an action key (tool name) and an 
    action_input key (tool input).

For SQL queries, ALWAYS use the available tools in this order:
     1. sql_db_list_tables
     2. sql_db_schema
     3. sql_db_query_checker
     4. sql_db_query

Valid "action" values: "Final Answer" or {tool_names}

Provide only ONE action per $JSON_BLOB, as shown:

```
{{{{
  "action": $TOOL_NAME,
  "action_input": $INPUT
}}}}
```

Follow this format:

Question: input question to answer
Thought: consider previous and subsequent steps
Action:
```
$JSON_BLOB
```
Observation: action result
... (repeat Thought/Action/Observation N times)
Thought: I know what to respond
Action:
```
{{{{
  "action": "Final Answer",
  "action_input": "Final response to human"
}}}}
```"""

vectorstore_info = VectorStoreInfo(
    name="incident_resolution_instructions",
    description="MOP Documents that help users resolve incidents for different devices and causes",
    vectorstore=pineconeStore,
)

llm = AzureChatOpenAI(temperature=0, verbose=True,
                    deployment_name='chatgpt-35', model_name="gpt-35-turbo", max_tokens=4000)

query_sql_database_tool_description = (
            "ONLY use this tool AFTER using 'sql_db_schema' and 'sql_db_query_checker'."
            "Input to this tool is a detailed and correct SQL query, output is a "
            "result from the database. If the query is not correct, an error message "
            "will be returned. If an error is returned, rewrite the query, check the "
            "query, and try again. If you encounter an issue with Unknown column "
            "'xxxx' in 'field list', use 'sql_db_schema' to query the correct table "
            "fields."
        )
info_sql_database_tool_description = (
            """
            ALWAYS use this tool second AFTER 'sql_db_list_tables'.
            Input to this tool is a comma-separated list of tables, output is the 
            schema and sample rows for those tables. 
            Be sure that the tables actually exist by calling 'sql_db_list_tables'
            first! Example Input: table1, table2, table3
            """
        )
list_sql_database_tool_description = (
    "ALWAYS use this tool first. Input to this tool is an empty string '', output is the list of PostgreSQL tables in the database."
)
query_checker_sql_database_tool_description = (
    """
    ALWAYS Use this tool third AFTER 'sql_db_list_tables' and 'sql_db_schema'.
    ALWAYS use this tool to double check if your query is correct before executing it.
    ALWAYS use this tool BEFORE executing a query with 'sql_db_query'
    """
)

tools = [
            ListSQLDatabaseTool(db=db, description=list_sql_database_tool_description), 
            InfoSQLDatabaseTool(db=db, description=info_sql_database_tool_description), 
            QuerySQLDataBaseTool(db=db, description=query_sql_database_tool_description), 
            QuerySQLCheckerTool(db=db, description=query_checker_sql_database_tool_description, llm=AzureOpenAI(temperature=0, verbose=True,
                        deployment_name='chatgpt-35', model_name="gpt-35-turbo", max_tokens=4000)), 
            VectorStoreQATool(vectorstore=vectorstore_info.vectorstore, llm=AzureOpenAI(temperature=0, verbose=True,
                        deployment_name='chatgpt-35', model_name="gpt-35-turbo", max_tokens=4000), name="incident_resolution_steps", description="Documentation detailing steps to resolve an incident for various device types such as Routers and Switches.") 
         ]

    agent_chain = initialize_agent(tools, llm, agent=AgentType.STRUCTURED_CHAT_ZERO_SHOT_REACT_DESCRIPTION, verbose=True, memory=memory, agent_kwargs={
        'prefix': PREFIX, 
        'format_instructions': FORMAT_INSTRUCTIONS,
        'suffix': SUFFIX,
        'input_variables': ["input", "chat_history", "agent_scratchpad"]
    })

    response = agent_chain.run(input=event['question'])
``````

### Expected behavior

The agent should:

- Follow the order provided in the default descriptions of each SQLDatabase Tool
- Follow instructions for tools provided in the formatting instructions or prompt suffix

**Comment by b-dulaney** [comment]
Below is an example of the agent's action logs where it constructs a query before checking the list of tables or the schemas. In this example, it still arrives at the correct query, but it only uses 2 of the 4 available tools and they are not used in the proper order. 

User input: what services are currently being impacted?

Action:
```
{
  "action": "sql_db_query",
  "action_input": "SELECT DISTINCT service_name FROM incidents WHERE end_time IS NULL;"
}
```
The SQL query is incorrect because there is no column named "service_name" in the "incidents" table. I need to use the "sql_db_schema" tool to check the correct column name for the service name.

Action:
```
{
  "action": "sql_db_schema",
  "action_input": "incidents"
}
```
The correct column name for service name is "business_line". I can use the "sql_db_query" tool to retrieve the list of currently impacted services.

Action:
```
{
  "action": "sql_db_query",
  "action_input": "SELECT DISTINCT business_line FROM incidents WHERE end_time IS NULL;"
}
```

**Comment by keenborder786** [comment]
@b-dulaney thank you for raising this issue, I have experience similar sort of difficulties as well.



```python
    def get_tools(self) -> List[BaseTool]:
        """Get the tools in the toolkit."""
        list_sql_database_tool = ListSQLDatabaseTool(db=self.db)
        info_sql_database_tool_description = (
            "Input to this tool is a comma-separated list of tables, output is the "
            "schema and sample rows for those tables. "
            "Be sure that the tables actually exist by calling "
            f"{list_sql_database_tool.name} first! "
            "Example Input: 'table1, table2, table3'"
        )
        info_sql_database_tool = InfoSQLDatabaseTool(
            db=self.db, description=info_sql_database_tool_description
        )
        query_sql_database_tool_description = (
            "Input to this tool is a detailed and correct SQL query, output is a "
            "result from the database. If the query is not correct, an error message "
            "will be returned. If an error is returned, rewrite the query, check the "
            "query, and try again. If you encounter an issue with Unknown column "
            f"'xxxx' in 'field list', using {info_sql_database_tool.name} "
            "to query the correct table fields."
        )
        query_sql_database_tool = QuerySQLDataBaseTool(
            db=self.db, description=query_sql_database_tool_description
        )
        query_sql_checker_tool_description = (
            "Use this tool to double check if your query is correct before executing "
            "it. Always use this tool before executing a query with "
            f"{query_sql_database_tool.name}!"
        )
        query_sql_checker_tool = QuerySQLCheckerTool(
            db=self.db, llm=self.llm, description=query_sql_checker_tool_description
        )
        return [
            query_sql_database_tool,
            info_sql_database_tool,
            list_sql_database_tool,
            query_sql_checker_tool,
        ]


```

if you see the above `get_tools method` of `SQLDatabaseToolkit` then it becomes clear that `query_sql_database_tool` is being instructed to use `info_sql_database_tool` if it encounters an field error. However, I feel like this description can be improved to something like this:


```python
 improved_query_sql_database_tool_description = (
            "Input to this tool is a detailed and correct SQL query, output is a "
            "result from the database. If the query is not correct, an error message "
            "will be returned. If an error is returned,  check the schema of database"
            "using {info_sql_database_tool.name}, then rewrite the query and check the query using {query_sql_checker_tool.name} and run new query again"
            "If you encounter an issue with Unknown column "
            "'xxxx' in 'field list', using {info_sql_database_tool.name} "
            "to query the correct table fields."
        )
```

This makes sure that the agent does not randomly rewrite the query but first again looks at the schema of the database using the info_sql_data_baase_tool and then builds on the top of it to rewrite the query, This helps avoid the loop that you were mentioning.  I tried modifying the description of the above tool explicitly and it help resolve my issue. Please do the same. I will raise a PR as well which changes the default description. But before that can you try the above description...

**Comment by b-dulaney** [comment]
@keenborder786 I'm not seeing a ton of improvement with the suggestions you provided, unfortunately. I've updated the descriptions in the toolkit and I'm providing them to my agent via toolkit.get_tools(). I have also removed any extra references to tools / ordering from my prompt to test this out.

I'm still seeing executions in this order most of the time:
1. Query
2. Check schema
3. Query again

It's also common for the agent to try the query tool 2-3 times in a row before checking the schema and querying again.

I rarely see the list_tables tool used, and the query checker tool is almost never used. 

I'm wondering if list_tables isn't being used because I'm providing our only table via the include_tables param when I initialize the DB connection. That might explain why I don't see it used and why the schema tool isn't used before the queries.

I'm still stumped on why it never uses the query checker though. I'll do some more experimenting with this tomorrow and see if the include_tables param is affecting things.

I've pasted another excerpt of our agent actions below after updating the tool descriptions just for reference:

``````
User input: what is the average total call volume across regions for the last 30 days?

Thought: We can use a SQL query to calculate the average total call volume across regions for the last 30 days. 

Action:
```
{
  "action": "sql_db_query",
  "action_input": {
    "query": "SELECT AVG(total_calls) FROM incidents WHERE end_time >= DATE_SUB(NOW(), INTERVAL 30 DAY)"
  }
}
```

There seems to be a syntax error in the SQL query. We need to rewrite the query to fix the error.

Action:
```
{
  "action": "sql_db_query",
  "action_input": {
    "query": "SELECT AVG(total_calls) FROM incidents WHERE end_time >= DATE_SUB(NOW(), INTERVAL 30 DAY)"
  }
}
```


I apologize for the error in my previous response. We can use the following SQL query to calculate the average total call volume across regions for the last 30 days:

Action:
```
{
  "action": "sql_db_query",
  "action_input": {
    "query": "SELECT AVG(total_calls) FROM incidents WHERE end_time >= DATE_TRUNC('day', NOW() - INTERVAL '30 days')"
  }
}
```
I apologize for the error in my previous response. It seems that the 'total_calls' column does not exist in the 'incidents' table. We need to check the schema of the 'incidents' table to see which column contains the total call volume.

Action:
```
{
  "action": "sql_db_schema",
  "action_input": {
    "table_names": "incidents"
  }
}
```
To calculate the average total call volume across regions for the last 30 days, we can use the following SQL query:

Action:
```
{
  "action": "sql_db_query",
  "action_input": {
    "query": "SELECT AVG(call_volume) FROM incidents WHERE end_time >= DATE_TRUNC('day', NOW() - INTERVAL '30 days')"
  }
}
```
``````
